src/instructlab/model/train.py

```diff
@@ -520,20 +580,340 @@ def get_files(directory: str, pattern: str) -> list[str]:
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    return ckpt_score
+
+
+def _evaluate_dir_of_checkpoints(
+    checkpoints_dir_path: str, eval_func: typing.Callable[..., float]
+) -> str:
+    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
+
+    # TODO: parallelize MMLU over available GPUs
+
+    results: list[typing.Tuple[float, str]] = []
+    directory = Path(checkpoints_dir_path)
+    for ckpt_path in directory.iterdir():
+        if ckpt_path.is_dir():
+            logger.debug(ckpt_path)
+            ckpt_score = eval_func(ckpt_path=ckpt_path)
             click.secho(
-                f"Error while executing training library {exc}",
+                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
                 fg="red",
+                bg="cyan",
             )
-            raise click.exceptions.Exit(1)
+            results.append((ckpt_score, ckpt_path))
+
+    # shape of: [tuple(score, path)]
+    ckpt_performance = sorted(results, reverse=True)
+
+    # (score, path)
+    return ckpt_performance[0]
+
+
+def _run_e2e_training(
+    ctx,
+    train_args: TrainingArgs,
+    torch_args: TorchrunArgs,
+    base_dir: str,
+    phase5_data: str,
+    phase5_num_epochs: typing.Optional[str],
+    phase5_samples_per_save: typing.Optional[str],
+    phase10_data: str,
+    phase10_num_epochs: typing.Optional[str],
+    phase10_samples_per_save: typing.Optional[str],
+    mtbench_judge: str,
+) -> None:
+
+    p5_ckpt_dir = os.path.join(base_dir, "phase5", "checkpoints")
+    p10_ckpt_dir = os.path.join(base_dir, "phase10", "checkpoints")
+
+    assert os.path.isdir(p5_ckpt_dir) and os.path.isdir(p10_ckpt_dir)
+    assert os.path.exists(phase5_data) and os.path.exists(phase10_data)
+
+    # cache these so we don't carry over phase5 override to phase10
+    DEFAULT_NUM_EPOCHS: int = train_args.num_epochs
+    DEFAULT_SAMPLES_PER_SAVE: int = train_args.save_samples
+
+    # -----------------------------------------------------------------------------
+    click.secho("Training Phase 1/2...", fg="cyan")
+    train_args.num_epochs = (
+        phase5_num_epochs if phase5_num_epochs else train_args.num_epochs
+    )
+    train_args.save_samples = (
+        phase5_samples_per_save if phase5_samples_per_save else train_args.save_samples
+    )
+    click.secho(f"Phase 1/2 training arguments:\n{pp.pformat(train_args)}", fg="cyan")
+
+    _e2e_training_phase(
+        train_args=train_args,
+        torch_args=torch_args,
+        data_path=phase5_data,
+        ckpt_dir=p5_ckpt_dir,
+        # model override not necessary because we expect model to come from ctx.params.model_path.
+    )
+
+    # -----------------------------------------------------------------------------
+    click.secho("MMLU evaluation for Phase 1...", fg="cyan")
+    hf_p5_ckpts = os.path.join(p5_ckpt_dir, "hf_format")
+
+    p5_best_ckpt: Path = None
```

Comment:
MyPy should technically catch this, but I recommend setting this to be `Path | None` so that you are forced to ensure `p5_best_ckpt` later on is `not None`.

Reply:
I changed it so that the function raises a RuntimeException if there's nothing to evaluate, so there will always be a p5_best_ckpt